Mathlib/RingTheory/Valuation/Discrete/Basic.lean

```diff
@@ -3,52 +3,77 @@ Copyright (c) 2025 María Inés de Frutos-Fernández, Filippo A. E. Nuccio. All
 import Mathlib.RingTheory.Valuation.Basic
 
 /-!
 # Discrete Valuations
 
-A valuation `v : A → ℤₘ₀` on a ring `A` is said to be a (normalized) discrete valuation if
-`ofAdd (-1 : ℤ)` belongs to the image of `v`. Note that valuations in Mathlib are multiplicative;
-if `a : A → ℤ ∪ {infty}` is the additive valuation associated to `v`, this is equivalent to asking
-that `1 : ℤ` belongs to the image of `a`.
+Given a linearly ordered commutative group with zero `Γ` such that `Γˣ` is nontrivial cyclic, a
+valuation `v : A → Γ` on a ring `A` is *discrete*, if `genLTOne Γˣ` belongs to the image, where
+`genLTOne Γˣ` is the generator of `Γˣ` that is `< 1`. When `Γ := ℤₘ₀` (defined in
+`Multiplicative.termℤₘ₀`), `genLTOne Γˣ = ofAdd (-1)` and the condition of being discrete is
+equivalent to asking that `ofAdd (-1 : ℤ)` belongs to the image, in turn equivalent to asking that
+`1 : ℤ` belongs to the image of the corresponding *additive* valuation.
+
 
 ## Main Definitions
-* `IsDiscrete`: We define a valuation to be discrete if it is `ℤₘ₀`-valued and `ofAdd (-1 : ℤ)`
-belongs to the image.
+* `IsDiscrete`: We define a `Γ`-valued valuation `v` to be discrete if `Γˣ` is cyclic and
+  `genLTOne Γˣ` belongs to the image of `v`
 
 ## TODO
-* Define (pre)uniformizers for nontrivial `ℤₘ₀`-valued valuations.
+* Define (pre)uniformizers for nontrivial discrete valuations.
 * Relate discrete valuations and discrete valuation rings.
-
 -/
 
 namespace Valuation
 
-open Function Multiplicative Set
+open Function Set LinearOrderedCommGroup
+
+variable {Γ : Type*} [LinearOrderedCommGroupWithZero Γ]
+
+variable {A : Type*} [Ring A] (v : Valuation A Γ)
 
-variable {A : Type*} [Ring A]
+/-- Given a linearly ordered commutative group with zero `Γ` such that `Γˣ` is
+nontrivial cyclic, a valuation `v : A → Γ` on a ring `A` is *discrete*, if
+`genLTOne Γˣ` belongs to the image. Note that the latter is equivalent to
+asking that `1 : ℤ` belongs to the image of the corresponding additive valuation. -/
+class IsDiscrete : Prop where
+  exists_generator_lt_one : ∃ (γ : Γˣ), Subgroup.zpowers γ = ⊤ ∧ γ < 1 ∧ ↑γ ∈ range v
 
-/-- A valuation `v` on a ring `A` is (normalized) discrete if it is `ℤₘ₀`-valued and
-  `ofAdd (-1 : ℤ)` belongs to the image. Note that the latter is equivalent to
-  asking that `1 : ℤ` belongs to the image of the corresponding additive valuation. -/
-class IsDiscrete (v : Valuation A ℤₘ₀) : Prop where
-  one_mem_range : (↑(ofAdd (-1 : ℤ)) : ℤₘ₀) ∈ range v
+lemma exists_generator_lt_one [IsDiscrete v] :
+  ∃ (γ : Γˣ), Subgroup.zpowers γ = ⊤ ∧ γ < 1 ∧ ↑γ ∈ range v := IsDiscrete.exists_generator_lt_one
+
+lemma IsDiscrete.cyclic_value_group [IsDiscrete v] : IsCyclic Γˣ := by
+  rw [isCyclic_iff_exists_zpowers_eq_top]
+  exact ⟨_, (v.exists_generator_lt_one.choose_spec).1⟩
+
+lemma IsDiscrete.nontrivial_value_group [IsDiscrete v] : Nontrivial Γˣ :=
+  ⟨1, v.exists_generator_lt_one.choose, ne_of_gt <| (v.exists_generator_lt_one.choose_spec).2.1⟩
```

Comment:
Do you think you should make `Valuation.generator [IsDiscrete v] : Gamma^*` to be `v.exists_generator_lt_one.choose` and then make some API for it like `lemma Valuation.generator_zpowers_eq_top`, `lemma Valuation.generator_lt_one` etc? Then you can avoid all this choose_spec.2.1 stuff. In short, will it be useful to actually be able to get to this generator easily with `v.generator`? It's unique so it feels like a good candidate for data.